JavaScript: Add support for E4X.

xiemaisi opened the pull request:
This PR adds extractor and library support for E4X and improves parsing of a few other Mozilla-specific language features. I'm still testing this, so it's not quite ready for merging yet, but I'm putting it up for review anyway.

Commit-by-commit review is strongly encouraged.

[Evaluation results](https://git.semmle.com/max/dist-compare-reports/blob/master/js/e4x/report.md) (internal link) show a lot of result differences, with a few syntax errors gone and some other alerts appearing in their place. Almost all changes are in test suites of JavaScript engines/parsers, and the ones I looked at mostly seem fine, except for the new unused-variable alerts in Lucifier; I'll take a closer look at those.

xiemaisi commented on javascript/extractor/src/com/semmle/jcorn/CustomParser.java:
```diff
@@ -320,4 +377,240 @@ protected Expression parseNew() {
+	 *     WildcardIdentifier
+	 *
+	 * QualifiedIdentifier :
+	 *     PropertySelector :: PropertySelector
```
This is what's causing the spurious alerts on Lucifier, since it causes an ambiguity with the proposed `::` function-bind operator.

asger-semmle commented on javascript/extractor/src/com/semmle/jcorn/CustomParser.java:
```diff
@@ -546,12 +546,21 @@ protected XMLAttributeSelector decoratorToAttributeSelector(Decorator d) {
+		// there is a lexical ambiguity between an XML processing instruction starting a
+		// chunk of E4X content and a Flow type annotation (both can start with `<?`)
+		// hence if we can't find the closing `?>` of a putative XML processing instruction
+		// we backtrack and try lexing as something else
```
Another idea for a dirty workaround: lookahead for the next `>`, and if that's preceded by `?` assume E4X otherwise Flow. A Flow type annotation will almost never contain `?>` and an XML processing instruction can't (?) contain a lone `>`, as it would need to be escaped as `&gt;`.

esben-semmle replied:
Another idea: can we commit to a parser mode for the rest of the compilation unit once we have observed the first evidence of a specific language variant? Then the lookahead problem will only occur if `<?` comes before all other evidence.

xiemaisi replied:
I've implemented the first proposal, restricting it to `<?xml`.

esben-semmle commented on javascript/extractor/src/com/semmle/jcorn/CustomParser.java:
```diff
@@ -196,8 +196,12 @@ protected Node parseLetExpression(Position startLoc, boolean maybeStatement) {
 		if (this.type == TokenType.braceL) {
-			if (!maybeStatement)
-				this.unexpected();
+			if (!maybeStatement) {
```
I can not find anything about let expressions on TC39, are they an E4X thing? If so, don't we need a e4x feature check around here?

xiemaisi replied:
They are a historic Mozilla-specific extension. We already had support for it, but this PR improves it. Both invocations of this method are guarded by `options.mozExtensions()` checks further up.

esben-semmle commented on javascript/extractor/src/com/semmle/jcorn/flow/FlowParser.java:
```diff
@@ -78,6 +77,8 @@ protected Token getTokenFromCode(int code) {
 		private Function<Token, Void> onToken;
 		private boolean inType, noAnonFunctionType;
 		private List<Token> tokens = new ArrayList<Token>();
+		private Function<SyntaxError, ?> onRecoverableError;
```
`Function<SyntaxError, Void>`, like for `onToken` above?

esben-semmle replied:
Alternatively, both could be changed to have type [`Consumer`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html).

xiemaisi replied:
Changing `?` to `Void` does not, unfortunately, work, since there is one place where we pass in a `Function<SyntaxError, Boolean>`.

xiemaisi commented on javascript/extractor/src/com/semmle/jcorn/CustomParser.java:
```diff
@@ -320,4 +343,170 @@ protected Expression parseNew() {
 		}
 		return res;
 	}
+
```
With hindsight, I consider splitting up the parser across separate classes a mistake. I don't currently have time to rectify that mistake, but I don't want to make it worse, either.

xiemaisi commented on javascript/extractor/src/com/semmle/jcorn/Parser.java:
```diff
@@ -1691,27 +1691,34 @@ protected ParenthesisedExpressions parseParenthesisedExpressions(DestructuringEr
 		return parenExprs;
 	}
 
+	protected boolean parseParenthesisedExpression(DestructuringErrors refDestructuringErrors,
```
I've added a comment explaining the meaning of the result.

xiemaisi commented on javascript/extractor/tests/mozilla/input/exprfns.js:
```diff
@@ -0,0 +1,7 @@
+
+function(b) {
+  if (b) {
+    function g(y)("other result")
```
Correct.

xiemaisi commented:
I've backed out support for parsing of qualified XML names (function-bind is more important) and addressed the other comments as detailed above. I've also added a db upgrade script and opened a corresponding internal PR, which is green.

The PR check failure is due to tab-space mixing, which I propose we address by autoformatting the whole extractor in a separate PR.

I would consider this ready for merging now (though I will do some more testing on customer code and put up fix PRs as necessary).
